### DISCUSSION TITLE
Where are the doc-strings???

### DISCUSSION THREAD

**Discussion by Hvass-Labs** [pydantic/pydantic#4116] [General]
I cannot seem to find the doc-strings for many of Pydantic's public functions and class-methods. See for example `BaseModel` where most class-methods are *completely* undocumented, and only a few of them have a short doc-string.

Are users supposed to read all the code to understand what it is doing?

I can see how extremely popular Pydantic is and how busy you are building it, and I'm sure everyone greatly appreciates the massive contribution you are making to the world (I am being serious here!), but please consider it this way:

For each hour you are saving by not documenting your code properly, every single user is probably wasting at least 2 hours (probably more like 10 hours, but let's be conservative in our estimates here). You probably have at least 100_000 developers using it, so that is at least 200_000 hours of wasted developer time, or around 100 normed work-years, for each hour you saved by scrimping on the docs.

Does it seem like a good investment that you save one hour but the community then has to waste 200_000 hours? Imagine what could be built in that amount of time!

Not only is it a tremendous waste of time if every single user has to read through your code to try and understand what it does, but it is also *very dangerous* because people will not know whether something is a bug or a feature, because the expected behaviour / semantics are not explained anywhere, so this can propagate bugs far down into the user-code before it gets discovered that it was actually a bug in your code.

Some open-source authors expect other people to come along and write their docs. But I think it is extremely rare that someone else comes along and documents other people's code. And it would probably take another person 10 times as long to properly understand and document your code, as it would take yourself to write good docs for your own code.

Finally, it sends a signal to the open source community that undocumented code is the accepted standard - and it even sends the signal that people who would like to have good docs must be kind of stupid, so people become afraid to even speak up that they don't understand the code.

I would suggest the following as minimum documentation:

- Each Python module / file should contain a header that explains in maybe 3-10 lines what the file contains and how it broadly connects with other components of the code-base.

- If all functions / classes / etc. in the file are similar, then the header should point to one of these whose code is especially well documented, so people can easily understand a "blue-print" for how the code works, in case they need to maintain or extend it.

- Each function / class / etc. should have a doc-string that clearly explains what it does and what all the arguments are. This also pops up as a tool-tip in the IDE, and it can be used to auto-generate doc web-pages using Sphinx.

- Don't expect anyone else to come along in the future and improve your docs. So you need to enforce a high standard for the docs and only accept code into your repo that is properly documented.

Altogether this would make it much easier to use your otherwise excellent software and contribute to its further development. So I hope this is something you will prioritize in the future.

Thanks!

**Comment by samuelcolvin** [comment]
Thanks for your suggestion. I am aware of this.

It will be improved in V2, I intend to generate API documentation from docstrings similar to dirty-equals and watchfiles.
